include/oneapi/dpl/experimental/kt/single_pass_scan.h

```diff
@@ -0,0 +1,333 @@
+    using _LookbackKernel =
+        oneapi::dpl::__par_backend_hetero::__internal::__kernel_name_provider<__lookback_kernel<_KernelName>>;
+
+    const ::std::size_t __n = __in_rng.size();
+
+    if (__n == 0)
+        return sycl::event{};
+
+    static_assert(_Inclusive, "Single-pass scan only available for inclusive scan");
+
+    assert("This device does not support 64-bit atomics" &&
+           (sizeof(_Type) < 64 || __queue.get_device().has(sycl::aspect::atomic64)));
+
+    // We need to process the input array by 2^30 chunks for 32-bit ints
+    constexpr ::std::size_t __chunk_size = 1ul << (sizeof(_Type) * 8 - 2);
+    const ::std::size_t __num_chunks = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __chunk_size);
+
+    auto __max_cu = __queue.get_device().template get_info<sycl::info::device::max_compute_units>();
+    constexpr ::std::size_t __workgroup_size = _KernelParam::workgroup_size;
+    constexpr ::std::size_t __data_per_workitem = _KernelParam::data_per_workitem;
+
+    // Avoid non_uniform n by padding up to a multiple of workgroup_size
+    std::uint32_t __elems_in_tile = __workgroup_size * __data_per_workitem;
+    ::std::size_t __num_wgs = oneapi::dpl::__internal::__dpl_ceiling_div(__n, __elems_in_tile);
+
+    constexpr int __status_flag_padding = SUBGROUP_SIZE;
+    std::uint32_t __status_flags_size = __num_wgs + 1 + __status_flag_padding;
+
+    _FlagStorageType* __status_flags = sycl::malloc_device<_FlagStorageType>(__status_flags_size, __queue);
+
+    auto __fill_event = __lookback_init_submitter<_LookbackInitKernel>{}(__queue, __status_flags, __status_flags_size,
+                                                                         __status_flag_padding, _FlagType::__oob_value);
+
+    sycl::event __prev_event = __fill_event;
+    for (int __chunk = 0; __chunk < __num_chunks; ++__chunk)
+    {
+        ::std::size_t __current_chunk_size = __chunk == __num_chunks - 1 ? __n % __chunk_size : __chunk_size;
+        ::std::size_t __current_num_wgs =
+            oneapi::dpl::__internal::__dpl_ceiling_div(__current_chunk_size, __elems_in_tile);
+        ::std::size_t __current_num_items = __current_num_wgs * __workgroup_size;
+
+        auto __event = __lookback_submitter<__data_per_workitem, __workgroup_size, _Type, _FlagType, _LookbackKernel>{}(
+            __queue, __prev_event, __in_rng, __out_rng, __binary_op, __n, __status_flags, __status_flags_size,
+            __current_num_items);
+        __prev_event = __event;
+    }
+
+    auto __free_event = __queue.submit([=](sycl::handler& __hdl) {
+        __hdl.depends_on(__prev_event);
+        __hdl.host_task([=]() { sycl::free(__status_flags, __queue); });
+    });
+
+    return __free_event;
+}
+
+} // namespace __impl
+
+template <typename _InRng, typename _OutRng, typename _BinaryOp, typename _KernelParam>
+sycl::event
+inclusive_scan(sycl::queue __queue, _InRng __in_rng, _OutRng __out_rng, _BinaryOp __binary_op, _KernelParam __param)
```

Comment:
Let's pass `__in_rng` and `__out_rng` by forward references to avoid copying. 

I am not totally sure on how to pass the resulting range, given that C++ range algorithms accept an iterator as result, e.g. https://en.cppreference.com/w/cpp/algorithm/ranges/copy and I do not the reason for it. It seems better to pass it as a range e.g. to support API like `inclusive_scan(queue, sycl_buffer_in, sycl_buffer_out, ...)` and to make sure if the output container has enough space.

Reply:
I changed the signature to take the ranges as forwarding refs.